### ISSUE TITLE
Trouble with inttypes.h when trying to compile a program that uses system inttypes.h on Windows 8 with Visual Studio 2013

### ISSUE THREAD

**Issue by naxxster** [libgit2/libgit2#3476]
When compile a program which uses system inttypes.h on Windows 8 with Visual Studio 2013,
I get these error messages.

```
1>C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\include\inttypes.h(34): error C2371: 'imaxdiv_t' : redefinition; different basic types
1>          c:\workspace\q4\external\sources\libgit2-0.22.2\include\git2\inttypes.h(54) : see declaration of 'imaxdiv_t'
1>C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\include\inttypes.h(36): error C2375: '_abs64' : redefinition; different linkage
1>          C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\include\stdlib.h(388) : see declaration of '_abs64'
1>C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\include\inttypes.h(37): error C2375: 'imaxdiv' : redefinition; different linkage
1>          c:\workspace\q4\external\sources\libgit2-0.22.2\include\git2\inttypes.h(284) : see declaration of 'imaxdiv'
```

**Comment by linquize** [comment]
What libgit2 version are you using?
Could you provide sample code?

Do you 
1. include `git2.h` before `inttypes.h`; or
2. include `inttypes.h` before `git2.h`?

**Comment by naxxster** [comment]
I'm using libgit2-0.22.2.
The order of inclusion doesn't matter for this problem. I tested both.
I actually don't include inttypes.h by my self. google breakpad include the file internally.
The version of google breakpad is 4da1cfbad248e846b6e0b4429b4eb639bb2d1970 from https://chromium.googlesource.com/breakpad/breakpad.

Here is my sample codes

main.cpp
```
#include "stdafx.h"

#include <git2.h>
#include <client/windows/handler/exception_handler.h>


int wmain(int argc, wchar_t* argv[])
{
	return 0;
}

```

stdafx.h
```
#pragma once

#include <SDKDDKVer.h>
```

build command from visual studio log files
```
"C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\CL.exe" /c /IC:\workspace\q4\msvs.v120\..\external\sources\breakpad\src /I"C:\workspace\q4\msvs.v120\..\external\sources\libgit2-0.22.2\include" /ZI /nologo /W4 /WX- /sdl /Od /Oy- /D _DEBUG /D WIN32 /D _CONSOLE /D _LIB /Gm /EHsc /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Yu"stdafx.h" /Fp"C:\workspace\q4\msvs.v120\..\_intermediate\libgit2_fail_sample\Debug\x86\libgit2_fail_sample.pch" /Fo"C:\workspace\q4\msvs.v120\..\_intermediate\libgit2_fail_sample\Debug\x86\\" /Fd"C:\workspace\q4\msvs.v120\..\_intermediate\libgit2_fail_sample\Debug\x86\vc120.pdb" /Gd /TP /analyze- /errorReport:prompt main.cpp
```

**Comment by ethomson** [comment]
> redefinition; different linkage

suggests that you're trying to include the libgit2 header files in your C++ application, and you've included them with a C++ linkage.

If you specify:

    extern "C" {
        #include <git2.h>
    }

does this work?

**Comment by naxxster** [comment]
The main cause is not the missing of the extern directive. But I will apply that for my project.

Thank you for the comments.

**Comment by ethomson** [comment]
I'm not sure that I understand - what is the main cause then?

**Comment by naxxster** [comment]
Since MSVC 2013, there is built-in "inttypes.h" header file. It is used by breakpad library which I'm using.

libgit2 provides another "inttypes.h" file for older MSVC. There was no built-in "inttypes.h" header file.

The two files defines same entities, so the compiler complains it's redefinitions.

**Comment by nkeck720** [comment]
Could you add a #ifndef statement of some sort? Perhaps like this

#ifndef INTTYPES_H
#define INTTYPES_H
#include <inttypes.h>
#endif

**Comment by linquize** [comment]
@nkeck720 Note that libgit2's `stdint.h` defines `_MSC_STDINT_H_` instead of `INTTYPES_H`, so that it cannot prevent duplicate include between built-in `stdint.h` and libgit2's `stdint.h`

**Comment by nkeck720** [comment]
Then why would it be causing compile issues is my question?

**Comment by nkeck720** [comment]
That is a stupid question,  I didn't read thoroughly.

**Comment by ethomson** [comment]
There's nothing wrong with defining a prototype multiple times.  You can have some symbol prototyped in multiple header files, provided that it's for the same function (ie the prototype is identical).  The problem is when the prototype is different, for example:

    int foobar(int one);
    void foobar(const char *a, const char *b);

Or because you're using different linkages.  Like one symbol is defined as C and the other in a C++ linkage, because name mangling.  These are different, and incompatible definitions in a C++ application:

    int foobar(int one);

    extern "C" {
        int foobar(int one);
    }

If you try to compile a program that does this, you will see the message that you posted above, about different linkages.

That's why `extern "C"` fixes this.  We merged the fix in #3479.  We'll backport it to the next maintenance release.  Until then, if you're including `inttypes.h` and `git2.h`, please continue to use the `extern "C"` workaround.

**Comment by naxxster** [comment]
The problem is `imaxdiv_t`, which is typedef for a struct.
I think `extern "C"` doesn't help in this case.

**Comment by ethomson** [comment]
If it was a redefinition, and not a linkage problem, then #3477 should have fixed it.

**Comment by englercj** [comment]
I am getting similar errors about redefinition of stdint types using Visual Studio 2015:

```
1>libgit2\libgit2-0.26.1\include\git2\stdint.h(105): error C2371: 'int_fast16_t': redefinition; different basic types
1>  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\stdint.h(36): note: see declaration of 'int_fast16_t'
1>libgit2\libgit2-0.26.1\include\git2\stdint.h(109): error C2371: 'uint_fast16_t': redefinition; different basic types
1>  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\stdint.h(40): note: see declaration of 'uint_fast16_t'
1>libgit2\libgit2-0.26.1\include\git2\stdint.h(132): warning C4005: 'INT8_MIN': macro redefinition
1>  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\stdint.h(48): note: see previous definition of 'INT8_MIN'
... etc...
```

Seems like the fix in #3477 only applied to `_MSC_VER < 1800` which 2015 is `1900`. Is there something else I should be doing to have this work when I include libgit2 into a c++ translation unit? Removing the version requirement check in common.h seemed to fix the issue for me.
